Add deck and note options.

holocronweaver opened the pull request:
Adds options to decks and notes.

* Most deck options are handled in a new `OptionsGroup` class which allows multiple decks to share options, reflecting Anki's very similar option groups. OptionsGroup is closely modeled after the deck options window in Anki.
* Decks now have a `description` and `creation_time`. The latter is important since card due dates are defined relative to deck creation time.
* Notes now have a few option attributes, all related to SRS in some way (due date, interval, etc.).
* Add some initial README documentation.
* Add more internal documentation about database fields 
* Add internal documentation about options.
* No tests yet.
* `akpg_col.py` now has field substitution by dict rather than array. This was necessary due to the large number of options needing to be inserted. I didn't use `format` though!

I tried to stick to the coding style that already existed. Hopefully my blasphemies were kept at a minimum. X-D

kerrickstaley commented:
Thanks for the PR! I don't have time to look at it right now but will try to get to it in the next two days.

kerrickstaley commented on genanki/__init__.py:
```diff
@@ -189,6 +196,29 @@ def __init__(self, model=None, fields=None, sort_field=None, tags=None, guid=Non
       # guid was defined as a property
       pass
 
+    ## Options ##
```
Per above comment, please move this out of the `Note` class and into the `Card` class.

kerrickstaley commented on genanki/__init__.py:
```diff
@@ -243,12 +276,67 @@ def _format_tags(self):
 
 
+class OptionsGroup:
+  def __init__(self, options_id=None, name=None):
```
I'd like to follow the pattern of allowing all arguments to be set via keyword arguments to `__init__`.

So you should write something like
```python
def __init__(
    self,
    options_id=None,
    options_group_name=None,
    max_time_per_answer=60,
    ...
    misc=0):
  self.options_id = options_id
  self.options_group_name = options_group_name
  ...
```

I realize this is extremely verbose, so you could also go this alternate route:
```python
class OptionsGroup:
  ATTRIBUTES_WITH_DEFAULTS = {
    'options_id': None,
    'options_group_name': None,
    'max_time_per_answer': 60,
    ...
  }

  def __init__(self, **kwargs):
    for attr, default in self.ATTRIBUTES_WITH_DEFAULTS.items():
      setattr(self, attr, kwargs.get(attr, default))
```
Either style is fine. However, the first style is advantageous in that it's easier for tools like IDEs to introspect. And it's pretty easy to automatically generate the code so it ends up being about the same amount of work.

holocronweaver replied:
Will do!

holocronweaver commented on genanki/__init__.py:
```diff
@@ -155,25 +156,31 @@ class Card:
+        self.ord,   # ord - which card template it corresponds to
+        now_ts,     # mod - modification time as seconds since Unix epoch
+        -1,         # usn - value of -1 indicates need to push to server
+        stage,      # type - 0=new, 1=learning, 2=review
```
Do you mean just within this function, or in the user-facing API as well?

holocronweaver replied:
Actually, I remember now why I didn't use `type` for a variable name: it's a Python reserved keyword. Perhaps we should make an exception in this case?

kerrickstaley replied:
It's better to closely match Anki's naming. I want `genanki` to be a lower-level library that more-or-less directly maps to the way Anki represents data. Plus, the user will probably have to end up learning the Anki names anyway, for example if they're digging into the `collections.anki2` SQLite file to debug something. Abstractions are leaky more often than people expect.

I's OK to use `type` as the name. If you need to access the real `type()` function, you can import the `builtins` module and call `builtins.type`. (Previously I also had an instance of this, calling a param `ord_` instead of `ord`. But I changed it because it's silly to convolute your external API for the convenience of internal code, and I wasn't using `ord()` anyway).

holocronweaver commented on genanki/__init__.py:
```diff
@@ -155,25 +156,31 @@ class Card:
+                    #       learning: due time as integer seconds since Unix epoch
+                    #       review: integer days relative to deck creation
+        interval,   # ivl - positive days, negative seconds
+        ease,       # factor - integer ease factor used by SRS, 2500 = 250%
```
Okydoke! I actually planned for and forgot to do this, heh.

holocronweaver commented:
I think I made all the requested changes. 

I moved the card options to a new `CardOptions` class. This simplifies sharing the same options among multiple cards, similar to what `OptionsGroup` does for decks,  so it's handy on several levels.

I also added `@property` for the aliases as you suggested, while leaving the 'real' internal names consistent with the Anki DB field names.

kerrickstaley commented on genanki/__init__.py:
```diff
@@ -151,29 +152,88 @@ def to_json(self, now_ts, deck_id):
+
 class Card:
-  def __init__(self, ord):
+  def __init__(self, ord, options=None):
```
With this `options` object, you can't directly set properties on the `Card`. For example, you can't do `my_card.type = 1`, you have to do `my_card.options.type = 1`, which seems like an unnecessary level of indirection and doesn't mesh with how other objects (`Note`s and `Deck`s) work. Can you make these direct properties of `Card` and get rid of the `CardOptions` class?

To achieve what you're trying to do with `CardOptions`, you can write a helper function (in your own code) like
```
def set_card_attributes(card, attrs):
    for a in attr:
        setattr(card, a, attrs[a])
```
and then write code like
```
my_attrs = {'type': 2, 'due': 1234}
for note in notes:
    for card in note.cards:
        set_card_attributes(card, my_attrs)
```

holocronweaver commented:
Okydoke, removed `CardOptions` and now `Card` has all the attributes. I left a utility function in `Note` to make the common case of setting all card options at once more concise.

aehlke commented:
Can this be merged still
